### PR TITLE
Add support for rolling update of stateful Elastigroups

### PR DESCRIPTION
**Pull request by a1exsh**
The respawn command detects stateful Elastigroup and recycles stateful
instances one at a time, instead of using the deploy API (that only works for
the groups without stateful integrations).

Optional flag `--batch-per-subnet` may be used to recycle all instance of one AZ/subnet at a time.

### CODE REVIEW

File: senza/respawn.py [hunk 1]
```diff
@@ -183,13 +183,66 @@ def respawn_elastigroup(
     Respawn all instances in the ElastiGroup.
     """
 
+    spotinst_account = elastigroup_api.get_spotinst_account_data(region, stack_name)
+
+    stateful_instances = elastigroup_api.get_stateful_instances(elastigroup_id, spotinst_account)
+    if stateful_instances:
+        respawn_stateful_elastigroup(elastigroup_id, stack_name, batch_size, stateful_instances, spotinst_account)
```
Comment (jmcs) [inline senza/respawn.py]: Codacy Issue found: [Line too long (114/100)](https://app.codacy.com/manual/zalando-bus/senza/pullRequest?prid=5259705)

**Comment by jmcs** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/zalando-bus/senza/pullRequest?prid=5259705&bid=17232599)

File: senza/respawn.py [hunk 2]
```diff
@@ -183,13 +183,66 @@ def respawn_elastigroup(
...
+    else:
+        respawn_stateless_elastigroup(elastigroup_id, stack_name, batch_size, spotinst_account)
+
+
+def respawn_stateful_elastigroup(
+    elastigroup_id: str, stack_name: str, batch_size: int, stateful_instances: list, spotinst_account, sleep_sec=10
```
Comment (jmcs) [inline senza/respawn.py]: Codacy Issue found: [Line too long (115/100)](https://app.codacy.com/manual/zalando-bus/senza/pullRequest?prid=5259705)

**Comment by a1exsh** [comment]
No *new* tests were broken by this PR :joy:

**Comment by jmcs** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 4
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/zalando-bus/senza/pullRequest?prid=5259705&bid=17232599)

File: senza/respawn.py [hunk 3]
```diff
@@ -183,13 +183,72 @@ def respawn_elastigroup(
     Respawn all instances in the ElastiGroup.
     """
 
+    spotinst_account = elastigroup_api.get_spotinst_account_data(region, stack_name)
+
+    stateful_instances = elastigroup_api.get_stateful_instances(elastigroup_id, spotinst_account)
+    if stateful_instances:
+        respawn_stateful_elastigroup(elastigroup_id, stack_name, batch_size, stateful_instances, spotinst_account)
```
Comment (jmcs) [inline senza/respawn.py]: Codacy Issue found: [Line too long (114/100)](https://app.codacy.com/manual/zalando-bus/senza/pullRequest?prid=5259705)

File: senza/respawn.py [hunk 4]
```diff
@@ -183,13 +183,72 @@ def respawn_elastigroup(
...
+
+def wait_for_stateful_elastigroup(elastigroup_id: str, spotinst_account, sleep_sec: float):
+    """
+    Waits for all stateful instances of the ElastiGroup to be in the ACTIVE state.
+    """
+    with Action("Waiting for all stateful instances to be in the ACTIVE state") as act:
+        while True:
+            time.sleep(sleep_sec)
+            act.progress()
+            stateful_instances = elastigroup_api.get_stateful_instances(elastigroup_id, spotinst_account)
```
Comment (jmcs) [inline senza/respawn.py]: Codacy Issue found: [Line too long (105/100)](https://app.codacy.com/manual/zalando-bus/senza/pullRequest?prid=5259705)

File: senza/respawn.py [hunk 5]
```diff
@@ -183,13 +183,72 @@ def respawn_elastigroup(
...
+        info(
+            "Recycling stateful instance {} ({} | {})".format(
+                instance['id'], instance['instanceId'], instance['privateIp']
+            )
+        )
+        elastigroup_api.recycle_stateful_instance(elastigroup_id, instance['id'], spotinst_account)
+        wait_for_stateful_elastigroup(elastigroup_id, spotinst_account, sleep_sec)
+
+
+def stateful_elastigroup_ready(stateful_instances: list):
```
Comment (jmcs) [inline senza/respawn.py]: Codacy Issue found: [Missing function docstring](https://app.codacy.com/manual/zalando-bus/senza/pullRequest?prid=5259705)

File: senza/respawn.py [hunk 6]
```diff
@@ -183,13 +183,72 @@ def respawn_elastigroup(
...
+
+    stateful_instances = elastigroup_api.get_stateful_instances(elastigroup_id, spotinst_account)
+    if stateful_instances:
+        respawn_stateful_elastigroup(elastigroup_id, stack_name, batch_size, stateful_instances, spotinst_account)
+    else:
+        respawn_stateless_elastigroup(elastigroup_id, stack_name, batch_size, spotinst_account)
+
+
+def respawn_stateful_elastigroup(
+    elastigroup_id: str, stack_name: str, batch_size: int, stateful_instances: list, spotinst_account, sleep_sec=10
```
Comment (jmcs) [inline senza/respawn.py]: Codacy Issue found: [Line too long (115/100)](https://app.codacy.com/manual/zalando-bus/senza/pullRequest?prid=5259705)

**Comment by jmcs** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 5
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/zalando-bus/senza/pullRequest?prid=5259705&bid=17232599)

File: senza/respawn.py [hunk 7]
```diff
@@ -183,13 +183,75 @@ def respawn_elastigroup(
...
+        info(
+            "Recycling stateful instance {} ({} | {})".format(
+                instance['id'], instance['instanceId'], instance['privateIp']
+            )
+        )
+        elastigroup_api.recycle_stateful_instance(elastigroup_id, instance['id'], spotinst_account)
+        wait_for_stateful_elastigroup(elastigroup_id, spotinst_account, sleep_sec)
+
+
+def stateful_elastigroup_ready(stateful_instances: list):
```
Comment (jmcs) [inline senza/respawn.py]: Codacy Issue found: [Missing function docstring](https://app.codacy.com/manual/zalando-bus/senza/pullRequest?prid=5259705)

File: senza/respawn.py [hunk 8]
```diff
@@ -183,13 +183,75 @@ def respawn_elastigroup(
...
+
+def wait_for_stateful_elastigroup(elastigroup_id: str, spotinst_account, sleep_sec: float):
+    """
+    Waits for all stateful instances of the ElastiGroup to be in the ACTIVE state.
+    """
+    with Action("Waiting for all stateful instances to be in the ACTIVE state") as act:
+        while True:
+            time.sleep(sleep_sec)
+            act.progress()
+            stateful_instances = elastigroup_api.get_stateful_instances(elastigroup_id, spotinst_account)
```
Comment (jmcs) [inline senza/respawn.py]: Codacy Issue found: [Line too long (105/100)](https://app.codacy.com/manual/zalando-bus/senza/pullRequest?prid=5259705)

File: senza/respawn.py [hunk 9]
```diff
@@ -183,13 +183,75 @@ def respawn_elastigroup(
...
+    """
+    Recycles stateful instances of the ElastiGroup one by one.
+    """
+
+    if batch_size is not None:
+        raise Exception("Batch size is not supported when respawning stateful ElastiGroups")
+
+    if not stateful_elastigroup_ready(stateful_instances):
+        raise Exception(
+            "Stateful ElastiGroup {} is not ready: some instances are not in the ACTIVE state".format(elastigroup_id)
```
Comment (jmcs) [inline senza/respawn.py]: Codacy Issue found: [Line too long (117/100)](https://app.codacy.com/manual/zalando-bus/senza/pullRequest?prid=5259705)

File: senza/respawn.py [hunk 10]
```diff
@@ -183,13 +183,75 @@ def respawn_elastigroup(
...
+
+    stateful_instances = elastigroup_api.get_stateful_instances(elastigroup_id, spotinst_account)
+    if stateful_instances:
+        respawn_stateful_elastigroup(elastigroup_id, stack_name, batch_size, stateful_instances, spotinst_account)
+    else:
+        respawn_stateless_elastigroup(elastigroup_id, stack_name, batch_size, spotinst_account)
+
+
+def respawn_stateful_elastigroup(
+    elastigroup_id: str, stack_name: str, batch_size: int, stateful_instances: list, spotinst_account, sleep_sec=10
```
Comment (jmcs) [inline senza/respawn.py]: Codacy Issue found: [Line too long (115/100)](https://app.codacy.com/manual/zalando-bus/senza/pullRequest?prid=5259705)

File: senza/respawn.py [hunk 11]
```diff
@@ -183,13 +183,75 @@ def respawn_elastigroup(
     Respawn all instances in the ElastiGroup.
     """
 
+    spotinst_account = elastigroup_api.get_spotinst_account_data(region, stack_name)
+
+    stateful_instances = elastigroup_api.get_stateful_instances(elastigroup_id, spotinst_account)
+    if stateful_instances:
+        respawn_stateful_elastigroup(elastigroup_id, stack_name, batch_size, stateful_instances, spotinst_account)
```
Comment (jmcs) [inline senza/respawn.py]: Codacy Issue found: [Line too long (114/100)](https://app.codacy.com/manual/zalando-bus/senza/pullRequest?prid=5259705)

**Comment by jmcs** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 11
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/zalando-bus/senza/pullRequest?prid=5259705&bid=17232599)

File: senza/respawn.py [hunk 12]
```diff
@@ -177,19 +179,117 @@ def respawn_auto_scaling_group(
...
+    with Action("Waiting for all stateful instances to be in the ACTIVE state") as act:
+        while True:
+            time.sleep(WAIT_FOR_ELASTIGROUP_SEC)
+            act.progress()
+            stateful_instances = elastigroup_api.get_stateful_instances(elastigroup_id, spotinst_account)
+            if stateful_elastigroup_ready(stateful_instances):
+                break
+
+
+def recycle_stateful_elastigroup_instance(elastigroup_id: str, instance: dict, spotinst_account):
```
Comment (jmcs) [inline senza/respawn.py]: Codacy Issue found: [Missing function docstring](https://app.codacy.com/manual/zalando-bus/senza/pullRequest?prid=5259705)

File: senza/respawn.py [hunk 13]
```diff
@@ -177,19 +179,117 @@ def respawn_auto_scaling_group(
...
+
+    info(
+        "Recycling {} stateful instances for ElastiGroup {} (ID {})".format(
+            len(stateful_instances), stack_name, elastigroup_id
+        )
+    )
+
+    if batch_per_subnet:
+        instances_by_subnet = stateful_elastigroup_instances_by_subnet(region, stateful_instances)
+        for subnet, subnet_instances in sorted(instances_by_subnet.items(), key=lambda item: item[0]):
```
Comment (jmcs) [inline senza/respawn.py]: Codacy Issue found: [Line too long (102/100)](https://app.codacy.com/manual/zalando-bus/senza/pullRequest?prid=5259705)

File: senza/respawn.py [hunk 14]
```diff
@@ -177,19 +179,117 @@ def respawn_auto_scaling_group(
...
+
+            wait_for_stateful_elastigroup(elastigroup_id, spotinst_account)
+
+    else:
+        for instance in sorted(stateful_instances, key=lambda i: i['privateIp']):
+            recycle_stateful_elastigroup_instance(elastigroup_id, instance, spotinst_account)
+            wait_for_stateful_elastigroup(elastigroup_id, spotinst_account)
+
+
+def stateful_elastigroup_instances_by_subnet(region: str, stateful_instances: list):
```
Comment (jmcs) [inline senza/respawn.py]: Codacy Issue found: [Missing function docstring](https://app.codacy.com/manual/zalando-bus/senza/pullRequest?prid=5259705)

File: senza/respawn.py [hunk 15]
```diff
@@ -177,19 +179,117 @@ def respawn_auto_scaling_group(
...
     """
 
+    spotinst_account = elastigroup_api.get_spotinst_account_data(region, stack_name)
+
+    stateful_instances = elastigroup_api.get_stateful_instances(elastigroup_id, spotinst_account)
+    if stateful_instances:
+        if batch_size is not None:
+            raise Exception("Batch size is not supported when respawning stateful ElastiGroups")
+
+        respawn_stateful_elastigroup(elastigroup_id, stack_name, region, batch_per_subnet, stateful_instances, spotinst_account)
```
Comment (jmcs) [inline senza/respawn.py]: Codacy Issue found: [Line too long (128/100)](https://app.codacy.com/manual/zalando-bus/senza/pullRequest?prid=5259705)

File: senza/respawn.py [hunk 16]
```diff
@@ -177,19 +179,117 @@ def respawn_auto_scaling_group(
...
+            subnet = "{} | {}".format(
+                i['Placement']['AvailabilityZone'], i['SubnetId']
+            )
+            instance = instances_by_ec2_id[i['InstanceId']]
+            instances_by_subnet[subnet].append(instance)
+
+    return instances_by_subnet
+
+
+def stateful_elastigroup_ready(stateful_instances: list):
```
Comment (jmcs) [inline senza/respawn.py]: Codacy Issue found: [Missing function docstring](https://app.codacy.com/manual/zalando-bus/senza/pullRequest?prid=5259705)

File: senza/respawn.py [hunk 17]
```diff
@@ -177,19 +179,117 @@ def respawn_auto_scaling_group(
...
+
+    stateful_instances = elastigroup_api.get_stateful_instances(elastigroup_id, spotinst_account)
+    if stateful_instances:
+        if batch_size is not None:
+            raise Exception("Batch size is not supported when respawning stateful ElastiGroups")
+
+        respawn_stateful_elastigroup(elastigroup_id, stack_name, region, batch_per_subnet, stateful_instances, spotinst_account)
+    else:
+        if batch_per_subnet:
+            raise Exception("Batch per subnet is not supported when respawning stateless ElastiGroups")
```
Comment (jmcs) [inline senza/respawn.py]: Codacy Issue found: [Line too long (103/100)](https://app.codacy.com/manual/zalando-bus/senza/pullRequest?prid=5259705)

File: senza/cli.py [hunk 18]
```diff
@@ -1954,7 +1959,7 @@ def respawn_instances(stack_ref, inplace, force, batch_size_percentage, region):
             )
         elif group["type"] == ELASTIGROUP_RESOURCE_TYPE:
             respawn.respawn_elastigroup(
-                group["resource_id"], group["stack_name"], region, batch_size_percentage
+                group["resource_id"], group["stack_name"], region, batch_size_percentage, batch_per_subnet
```
Comment (jmcs) [inline senza/cli.py]: Codacy Issue found: [Line too long (106/100)](https://app.codacy.com/manual/zalando-bus/senza/pullRequest?prid=5259705)

File: senza/respawn.py [hunk 19]
```diff
@@ -177,19 +179,117 @@ def respawn_auto_scaling_group(
...
+
+def wait_for_stateful_elastigroup(elastigroup_id: str, spotinst_account):
+    """
+    Waits for all stateful instances of the ElastiGroup to be in the ACTIVE state.
+    """
+    with Action("Waiting for all stateful instances to be in the ACTIVE state") as act:
+        while True:
+            time.sleep(WAIT_FOR_ELASTIGROUP_SEC)
+            act.progress()
+            stateful_instances = elastigroup_api.get_stateful_instances(elastigroup_id, spotinst_account)
```
Comment (jmcs) [inline senza/respawn.py]: Codacy Issue found: [Line too long (105/100)](https://app.codacy.com/manual/zalando-bus/senza/pullRequest?prid=5259705)

File: senza/respawn.py [hunk 20]
```diff
@@ -177,19 +179,117 @@ def respawn_auto_scaling_group(
...
+def respawn_stateful_elastigroup(
+    elastigroup_id: str, stack_name: str, region: str, batch_per_subnet: bool, stateful_instances: list, spotinst_account
+):
+    """
+    Recycles stateful instances of the ElastiGroup.
+    """
+
+    if not stateful_elastigroup_ready(stateful_instances):
+        raise Exception(
+            "Stateful ElastiGroup {} is not ready: some instances are not in the ACTIVE state".format(elastigroup_id)
```
Comment (jmcs) [inline senza/respawn.py]: Codacy Issue found: [Line too long (117/100)](https://app.codacy.com/manual/zalando-bus/senza/pullRequest?prid=5259705)

File: senza/respawn.py [hunk 21]
```diff
@@ -177,19 +179,117 @@ def respawn_auto_scaling_group(
...
+        respawn_stateful_elastigroup(elastigroup_id, stack_name, region, batch_per_subnet, stateful_instances, spotinst_account)
+    else:
+        if batch_per_subnet:
+            raise Exception("Batch per subnet is not supported when respawning stateless ElastiGroups")
+
+        respawn_stateless_elastigroup(elastigroup_id, stack_name, batch_size, spotinst_account)
+
+
+def respawn_stateful_elastigroup(
+    elastigroup_id: str, stack_name: str, region: str, batch_per_subnet: bool, stateful_instances: list, spotinst_account
```
Comment (jmcs) [inline senza/respawn.py]: Codacy Issue found: [Line too long (121/100)](https://app.codacy.com/manual/zalando-bus/senza/pullRequest?prid=5259705)

File: senza/cli.py [hunk 22]
```diff
@@ -1933,11 +1933,16 @@ def patch_spotinst_elastigroup(properties, elastigroup_id, region, stack_name):
     metavar="PERCENTAGE",
     type=click.IntRange(0, 100, clamp=True),
     help="Percentage (int value) of the ElastiGroup cluster that is respawned in each step."
-    " Valid only for ElastiGroups. The default value for this of 20.",
+    " Valid only for stateless ElastiGroups. The default value for this of 20.",
+)
+@click.option(
+    "--batch-per-subnet",
+    is_flag=True,
+    help="Recycle ElastiGroup instances in batches per subnet. Valid only for stateful ElastiGroups.",
```
Comment (jmcs) [inline senza/cli.py]: Codacy Issue found: [Line too long (102/100)](https://app.codacy.com/manual/zalando-bus/senza/pullRequest?prid=5259705)

**Comment by jmcs** [comment]
@a1exsh do you still want this change merged? If so please rebase the branch. The unit tests are fixed on master.

**Comment by a1exsh** [comment]
@jmcs yes, we use the patched version and it saves us a lot of hassle.  Thanks for fixing the tests! :)

**Comment by jmcs** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 11
           
```


See the complete overview on [Codacy](https://app.codacy.com/manual/zalando-bus/senza/pullRequest?prid=5259705&bid=17232599)

File: senza/respawn.py (hunk 15 above)
Comment (jmcs) [inline senza/respawn.py:195]: Codacy found an issue: [Line too long (128/100)](https://app.codacy.com/manual/zalando-bus/senza/pullRequest?prid=5259705)

File: senza/respawn.py (hunk 20 above)
Comment (jmcs) [inline senza/respawn.py:212]: Codacy found an issue: [Line too long (117/100)](https://app.codacy.com/manual/zalando-bus/senza/pullRequest?prid=5259705)

File: senza/respawn.py (hunk 12 above)
Comment (jmcs) [inline senza/respawn.py:272]: Codacy found an issue: [Missing function docstring](https://app.codacy.com/manual/zalando-bus/senza/pullRequest?prid=5259705)

File: senza/respawn.py (hunk 13 above)
Comment (jmcs) [inline senza/respawn.py:223]: Codacy found an issue: [Line too long (102/100)](https://app.codacy.com/manual/zalando-bus/senza/pullRequest?prid=5259705)

File: senza/respawn.py (hunk 14 above)
Comment (jmcs) [inline senza/respawn.py:238]: Codacy found an issue: [Missing function docstring](https://app.codacy.com/manual/zalando-bus/senza/pullRequest?prid=5259705)

File: senza/cli.py (hunk 22 above)
Comment (jmcs) [inline senza/cli.py:1941]: Codacy found an issue: [Line too long (102/100)](https://app.codacy.com/manual/zalando-bus/senza/pullRequest?prid=5259705)

File: senza/respawn.py (hunk 17 above)
Comment (jmcs) [inline senza/respawn.py:198]: Codacy found an issue: [Line too long (103/100)](https://app.codacy.com/manual/zalando-bus/senza/pullRequest?prid=5259705)

File: senza/respawn.py (hunk 21 above)
Comment (jmcs) [inline senza/respawn.py:204]: Codacy found an issue: [Line too long (121/100)](https://app.codacy.com/manual/zalando-bus/senza/pullRequest?prid=5259705)

File: senza/respawn.py (hunk 16 above)
Comment (jmcs) [inline senza/respawn.py:255]: Codacy found an issue: [Missing function docstring](https://app.codacy.com/manual/zalando-bus/senza/pullRequest?prid=5259705)

File: senza/cli.py (hunk 18 above)
Comment (jmcs) [inline senza/cli.py:1962]: Codacy found an issue: [Line too long (106/100)](https://app.codacy.com/manual/zalando-bus/senza/pullRequest?prid=5259705)

File: senza/respawn.py (hunk 19 above)
Comment (jmcs) [inline senza/respawn.py:267]: Codacy found an issue: [Line too long (105/100)](https://app.codacy.com/manual/zalando-bus/senza/pullRequest?prid=5259705)

**Comment by a1exsh** [comment]
:+1:

**Comment by jmcs** [comment]
:+1:

**Pull request merged**